### PR TITLE
Add exercises back to episode 3

### LINKED ISSUE #183: New exercises for chapter 3
Since we converted the exercises of chapter 3 to main material we don't have as many exercises left.

I propose the first step is to extract useful nuggets from the old exercises and adding them back in.

### PR DESCRIPTION
**Pull request by dafnevk**
This PR adds back useful parts of the removed exercises in episode 3. Fixes #183.

### CODE REVIEW

File: _episodes/03-monitor-the-model.md [hunk 1]
```diff
@@ -366,6 +382,17 @@ axes[1].set_ylabel("true sunshine hours")
 {: .language-python}
 ![Scatter plot to evaluate training and test set](../fig/03_regression_training_test_comparison.png)
 
+> ## Exercise: Reflecting on our results
+> 1. Is the performance of the model as you expected (or better/worse)?
+> 2. Is there a noteable difference between training set and test set? And if so, any idea why?
+>
+> > ## Solution
+> >  
+> > While the performance on the train set seems reasonable, the performance on the test set is much worse.
```
Comment (psteinb) [inline _episodes/03-monitor-the-model.md:391]: What hint to the cause of this would you give?

File: _episodes/03-monitor-the-model.md [hunk 2]
```diff
@@ -486,79 +518,80 @@ That is a clear signature of overfitting.
...
-history_df = pd.DataFrame.from_dict(history.history)
-sns.lineplot(data=history_df[['root_mean_squared_error', 'val_root_mean_squared_error']])
-plt.xlabel("epochs")
-plt.ylabel("RMSE")
-~~~
-{:.language-python}
-
-![Output of plotting sample](../fig/03_training_history_3_rmse_smaller_model.png)
-
+> ## Try to reduce the degree of overfitting by lowering the number of parameters
```
Comment (psteinb) [inline _episodes/03-monitor-the-model.md:521]: I personally think that the content of this exercise reads more like teaching content. I would object to make it an exercise. Can you comment on why you took that step?

File: _episodes/03-monitor-the-model.md [hunk 3]
```diff
@@ -198,9 +198,23 @@ Setting the `random_state` to `0` is a short-hand at this point. Note however, t
...
+Now we want to work on a *regression task*, thus not predicting a class label (or integer number) for a datapoint. In regression, we like to predict one (and sometimes many) values of a feature. This is typically a floating point number.
 
-In our example we want to predict the sunshine hours in Basel (or any other place in the dataset) for tomorrow based on the weather data of all 18 locations today. `BASEL_sunshine` is a floating point value (i.e. `float64`). The network should hence output a single float value which is why the last layer of our network will only consist of a single node. 
+> ## Exercise: Architecture of the network
+> As we want to design a neural network architecture for a regression task,
+> see if you can first come up with the answers to the following questions:
+> 1. What must be the dimension of our input layer?
+> 2. How would our output layer look like? What about the activation function? Tip: Remember that the activation function in our previous classification network scaled the outputs between 0 and 1.
```
[the remark this answers is not in the document]
Comment (dafnevk) [reply _episodes/03-monitor-the-model.md:207]: The idea is to let the learners think about this a bit before giving them the solution on how to do this directly. But you're right that it's a tricky question, especially the activation function. Would you have suggestions on how to rephrase the question so that we still trigger this thought process, but without making it too difficult?

File: _episodes/03-monitor-the-model.md (hunk 1 above)
Comment (dafnevk) [reply _episodes/03-monitor-the-model.md:391]: We're diving into that in the lesson content that follows of course, but indeed maybe we can give a hint here already.

File: _episodes/03-monitor-the-model.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (psteinb) [reply _episodes/03-monitor-the-model.md:207]: That is a good question. the challenge is, that `softmax` involves quite some math from the point of view of a learner. Here, we could argue, that this math is simply not needed first. This way learners are reassured that they need to consider something else. Then, the concept of the activation function transforming the outputs is important as it has direct consequences onto the variable being regressed.

How about we add a concrete hint:
```suggestion
> 2. We want to to output the prediction of a single number. The output layer of the NN hence cannot be the same as for the classification task earlier. This is because the `softmax` activation being used had a concrete meaning with respect to the class labels which is not needed here. What output layer design would you choose for regression?  
Hint: A layer with `relu` activation, with `sigmoid` activation or no activation at all? 
```

**Comment by dafnevk** [comment]
The excercise on splitting the train and test set could be removed, see #198

File: _episodes/03-monitor-the-model.md (hunk 1 above)
Comment (psteinb) [reply _episodes/03-monitor-the-model.md:392]: Ok, so this is just an observation.

File: _episodes/03-monitor-the-model.md [hunk 4]
```diff
@@ -486,79 +527,80 @@ That is a clear signature of overfitting.
...
+> > history_df = pd.DataFrame.from_dict(history.history)
+> > sns.lineplot(data=history_df[['root_mean_squared_error', 'val_root_mean_squared_error']])
+> > plt.xlabel("epochs")
+> > plt.ylabel("RMSE")
+> > ~~~
+> > {:.language-python}
+> >
+> > ![Output of plotting sample](../fig/03_training_history_3_rmse_smaller_model.png)
+> >
+> > There is obviously no single correct solution here. But you will have noticed that the number of nodes can be reduced quiet a bit!
```
Comment (psteinb) [inline _episodes/03-monitor-the-model.md:598]:
```suggestion
> > There is no single correct solution here. But you will have noticed that the number of nodes can be reduced quiet a bit!
```

File: _episodes/03-monitor-the-model.md (hunk 2 above)
Comment (psteinb) [reply _episodes/03-monitor-the-model.md:530]: You could use a different part of the dataset for the exercise.

File: _episodes/03-monitor-the-model.md [hunk 5]
```diff
@@ -431,6 +460,18 @@ NN RMSE: 4.05, baseline RMSE: 3.88
...
 Judging from the numbers alone, our neural network preduction would be performing worse than the baseline.
 
+> ## Exercise: Baseline
+> Looking at this baseline: Would you consider this a simple or a hard problem to solve?
+>
+> > ## Solution
+> >  
+> > This really depends on your definition of hard! The baseline gives a more accurate prediction than just
+> > randomly predicting a number, so the problem is not impossible. However, if no model is going to
+> > perform better than the baseline, the problem may still be very hard.
```
Comment (psteinb) [inline _episodes/03-monitor-the-model.md:470]: The last sentence of this solution appears quite confusing to me. Just a suggestion:

```suggestion
> > This really depends on your definition of hard! The baseline gives a more accurate prediction than just
> > randomly predicting a number, so the problem is not impossible to solve with machine learning. However, given the structure of the data and our expectations with respect to quality of prediction, it may remain hard to find a good algorithm which exceeds our baseline by orders of magnitude. 
```

File: _episodes/03-monitor-the-model.md [hunk 6]
```diff
@@ -431,6 +458,11 @@ NN RMSE: 4.05, baseline RMSE: 3.88
 
 Judging from the numbers alone, our neural network preduction would be performing worse than the baseline.
 
+> ## Exercise: Baseline
+> Looking at this baseline: Would you consider this a simple or a hard problem to solve?
```
[the remark this answers is not in the document]
Comment (dafnevk) [reply _episodes/03-monitor-the-model.md:464]: Do you think the question itself needs an additional hint?

File: _episodes/03-monitor-the-model.md (hunk 2 above)
Comment (dafnevk) [reply _episodes/03-monitor-the-model.md:529]: There is indeed an exercise like that already added in PR #196 . But that's at the very end, as a wrap up exercise, I like that this exercise is about halfway through. My preference is to leave it in as an exercise. I think the exercise can be safely skipped if needed, then there's just the mention of the fact that reducing parameters could work.
